Introduce Sinks and deprecate concrete processors

simonbasle opened the pull request:
Last debated and designed in #2179 but also relates to #1169, #1296, #1172, #1170

Fixes #2179
Fixes #2203

spencergibb commented on reactor-core/src/main/java/reactor/core/publisher/DirectProcessor.java:
```diff
@@ -77,7 +78,9 @@
  * </p>
  *
  * @param <T> the input and output value type
+ * @deprecated Prefer clear cut usage of either {@link Processors} or {@link Sinks}, to be removed in 3.5
```
Can these wait until a major to be removed?

simonbasle replied:
can you expand on your concern ? if we follow the history of Reactor (2.x > 3.0 > 3.1 > 3.2 and now 3.3), a "major" would be a significant event that would probably mean a complete re-architecture, to the point that the whole ecosystem around Reactor would probably need heavy re-architecturing too, or break (see Reactor 2 vs Reactor 3). The only other drives for a Reactor 4 that I could think about are a change in the Java baseline or the advent of a Reactive Streams v2. In any case, very unlikely at this point.

spencergibb replied:
deprecating one release and removing the next just breaks backward compatibility quickly. With boot moving to a 6-month cadence for minors, they are deprecating, then waiting an extra release to remove. So deprecate in x.1, do nothing in x.2 and remove in x.3.

simonbasle replied:
I see. But unlike Boot, we haven't switched to a 6 month cadence. So deprecating these now would mean that the concrete classes would get hidden in a year or so. Initially, there was an issue to directly remove them without the deprecation, but I agree this would be overkill. Note that `TopicProcessor` and `WorkQueueProcessor` are slated to be removed though (since deprecated in 3.3.0)

spencergibb replied:
:+1: got it. thanks for taking the time to explain.

smaldini commented:
Did a first pass here are my comments:
- Don't like having a `Processors` vs just the new `Sinks`. Adding a new top level class would signal continuity in support for `Processor` and still a double entry point for a user that can just use sinks. I understand we can't provide an easy bridge with `SinkFlux.toProcessor()` without allocating a serialized sink, and that would hurt some operator perfs.

- i like the idea of where `Processors` and `more()` was going with an allocation free guided API. Could be used to differentiate flux vs mono sinks. Also, given my comment above the concept could be applied to Sinks with `unsafe()` to produce non serialized processors. Said differently, that means all sinks would be serialized by default, in fact I'm not sure it's the sink itself that needs serialization or a processor (via delegate) so toProcessor() comes serialized as well.

- `FluxSink` vs `SinkFlux` is going to be confusing. I've suggested in the past to bring back an old naming from history, `Broadcaster`, the intent being sort of pub-sub.

- Sinks factories could reuse more of the existing flux vocabulary: `onBackpressureDrop` etc maybe ? 

- Should sinks share some `ConnectableFlux` concept ? 

I'd conclude with this final thought: Ultimately processors are all-in-one series of behaviors available in fluxes and I'm excited to see a long-time dream of rationalization happening with them. Today we can write pub-sub bridges using processors with and without sinks or `Flux.create`. All of these constructs ask the user what to do with backpressure, completely overlapping with existing `Flux.onBackpressureXXX` operator. In addition, `EmitterProcessor` uniquely overlaps with `flux.share()` and `UnicastProcessor` looks like an `Operators` util more than anything. A good place to be first is to get rid of public exposure of a `Processor`. Because processors are N behaviors in one, they are pretty much a technical optimization (allocation wise) over a longer flux definition. I'll need to iterate a bit more on that just wanted to share my first impression.

cc @rstoyanchev

simonbasle commented:
@smaldini thanks for the comments.

> FluxSink vs SinkFlux is going to be confusing. I've suggested in the past to bring back an old naming from history, Broadcaster, the intent being sort of pub-sub.

First and foremost, the sink naming is definitely going to evolve. I thought about `Broadcaster`, but then I figured having a sink that enforces unicast / single subscriber could be a real use case ?

Maybe we can solve that by having a `.unicast()` operator, effectively allowing turning any publisher into a unicast one that rejects more than one active subscription?? But that said, in the case of the sink it would introduce overhead (due to the fact that the underlying implementation would be multicast, currently a `ReplayProcessor`).

The trouble I have with the existing `FluxSink` interface is that it stems from operators like handle or create, which makes these sinks "unicast" -> one is created for each `Subscriber`. As a consequence it makes sense for them to have subscription-like methods (`onRequest`, `onCancel`, `requestedFromDownstream`, etc...) which would make less sense for the new flavor, the *standalone* sink.

Ie. an operator sink is more like a Subscription to subscribers, whereas a standalone sink is more like a shared hot Publisher to subscribers.

> Sinks factories could reuse more of the existing flux vocabulary: onBackpressureDrop etc maybe ?

Again, naming to evolve. But yeah, I can see this mapping to `onBackpressureXxx` terminology. Maybe if we make a builder for sinks?

I initially refrained from referring to backpressure because the concern here is about what happens when you push data and there is NO subscriber. Once there are subscribers, the sinks would honor backpressure (like the `ReplayProcessor` does).

dfeist commented:
>  I thought about Broadcaster, but then I figured having a sink that enforces unicast / single subscriber could be a real use case?
'Sink` is definitely preferred over `Broadcaster` IMO.  The term `sink` (or less often `source`) is often used in message processing and is cardinality agnostic.  Whereas "broadcast" implies it's to `n` subscribers, which isn't always the case

>Maybe we can solve that by having a .unicast()
Wouldn't you just choose to use a `Sink` implementation that enforces this?  The `.unicast()` type operators are good but have to be careful not to overuse them.

Sinks need to handle this use case: https://github.com/reactor/reactor-core/issues/1037

simonbasle commented:
> Sinks need to handle this use case: #1037

@dfeist I wish there wasn't this additional consideration that piles up with all the others 😭 `tryNext` support (or planning for future support) is actually very constraining because it implies direct access to the various subscribers' requests (in order to reject the value if at least once is at 0). Which kinda means duplicating multicasting code (either processors or `ConnectableFlux` sources) for the benefit of that sole feature. Which means additional maintenance burden, when we'd like to reduce that burden with this change.

I'm pretty depressed at the state of the whole affair to be honest...
There no clear path forward, two many axis to consider:

 - standalone use vs in-operator / per-subscriber use
   - misalignment of the current `FluxSink` API that bears methods that are very subscriber-oriented (`requestedFromDownstream`, `onCancel`, etc...)
   - misalignment of `Mono` vs `Flux` in terms of multicasting operators
 - behavior when calling `next` BEFORE the first `Subscriber` comes in
 - behavior when calling `complete`/`error` BEFORE the first `Subscriber` comes in
 - multicast or not
 - replay or not
 - behavior once subscribed to honor backpressure
 - `tryNext` which implies knowledged of the various downstream requests in case of multicast
 - ability to downgrade from serialized sink to raw sink
 - ability to downgrade from sink to processor

simonbasle commented:
@smaldini @dfeist I reworked the PR:

 - for naming, I settled on adding `ScalarSink` and `SequenceSink` as a base interface.
    - `FluxSink` and `MonoSink` can thus be kept as-is (avoiding painful migration on that side of things, eg. `Flux.create` is untouched)
    - adding `StandaloneFluxSink` and `StandaloneMonoSink` as a parallel set of interfaces, because the `onCancel`/etc... methods of `FluxSink` don't make sense in a standalone context
 - `MonoProcessor` is exposed as the `trigger()` sink
 - After trying out a Flux-to-Mono kind of sink, I abandoned the idea because I'm not sure there's a clear-cut case for it (yet)
 - kept `Processors` as a way to ultimately offer `FluxProcessor`s without exposing the concrete classes. I don't think keeping the factory methods on the deprecated concrete class will help with the migration, making it too painful for users that disallow deprecated classes usage.

Time is running short (M1 is on Monday) so I feel I had to cut corners, but eh.

robotmrv commented on reactor-core/src/main/java/reactor/core/publisher/Sinks.java:
```diff
@@ -0,0 +1,262 @@
+
+		@Override
+		public Flux<T> asFlux() {
+			return processor;
```
Since `asFlux()` exposes `Flux<T>` but not concrete `Processor` `onEachOperator` hook could be applied here.  
Since `FluxProcessor`s are `ASYNC` `Publisher`s it could help with `Thread` context propagations

simonbasle replied:
the processor is only fed via manual interactions, so I'm not sure how that would help eg. if the user decides to do a `new Thread(() -> sink.next(1)).start()`...

robotmrv replied:
it would help downstream consumers as processor could be decorated and values would be emitted within context for them.  
here the example of the problem   
```java
//add hook that propagates value from Reactor `Context` to the `ThreadLocal`
....
String testValue = "testValue";
setTL(testValue);
MonoProcessor<Integer> processor = MonoProcessor.create();
Mono.delay(Duration.ofMillis(500))
        .doOnNext(it -> processor.onNext(1))
        .subscribe();

String value = processor
//        .doOnNext(it -> {}) // works only with dummy operator as it decorates downstream
        .map(it -> getValueFromTL())// value was not propagated because hook was not applied to the source
        .block();
Assertions.assertThat(value).isEqualTo(testValue);
 
```

simonbasle replied:
I need to look deeper into the consequences of that, not sure applying the hooks to the processor will work naturally. plus currently there is no such thing with processors. can you chime back in once M1 is released (eg creating an issue for M2) ?

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/FluxSink.java:
```diff
@@ -31,12 +31,10 @@
-     * @see Subscriber#onComplete()
-     */
-    void complete();
+	@Override
```
shouldn't it inherit the javadocs?

simonbasle replied:
the javadoc is automatically copied, inheritDoc is only if you want to supplement it with additional content while including the `super` content

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/Processors.java:
```diff
@@ -0,0 +1,143 @@
+
+
+	@SuppressWarnings("deprecation")
+	public static final <T> FluxProcessor<T, T> unicast() {
```
FYI one of the reasons why I was choosing concrete type over `FluxProcessor` is because having to specify the type argument twice was a bit cumbersome.

As this is a new API, perhaps we can do something about it? especially given that now it is *always* `<T, T>`

simonbasle replied:
the number of new interfaces / classes was a concern in the original proposal back then, but there's probably little way around the current split into 4 sink interfaces (plus 2 additional utility classes).

we could improve that `<T,T>` situation but it means adding `FluxIdentityProcessor<T>` at least... Maybe `Processors` can be turned into both interface and factory.

 It raises another issue, which is the misalignment between `FluxProcessor<A,B>` and `MonoProcessor<A>` (mono processor has a single concrete implementation, and thus doesn't help with representing processing stages). That'll be a concern for either M2 or 3.5 now :/

simonbasle replied:
actually it cannot be melded with `Processors`, the `FluxProcessor` being an abstract class (due to the need to extend `Flux`)... furthermore `replay` factory method on `Processors` would clash with `Flux#replay()`, and the factory methods would be lost in the noise of `Flux` API.

Instead I went ahead and added an abstract intermediate `FluxIdentityProcessor<T> extends FluxProcessor<T, T>`

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/Sinks.java:
```diff
@@ -0,0 +1,262 @@
+	 * </ul>
+	 */
+	@SuppressWarnings("deprecation")
+	public static <T> StandaloneFluxSink<T> multicastNoWarmup() {
```
"warmup" sounds like a new term added. And, TBH, I struggle to understand it (even after reading the javadoc) 😅 
Could you please help me understanding the concept?

simonbasle replied:
yeah I'm having trouble finding a better term. It's the way some processors will "remember" upstream `onNext` (or in our case `sink.next`) that come in BEFORE the first `Subscriber` subscribes. But then the second `Subscriber` would only see live pushes.

```
\ denotes a subscription
<-warmup------><---------hot ---------->

--(1)---(2)----\---------\----(3)------>   sink
              -\(1)(2)--------(3)------>   sub1
                        -\----(3)------>   sub2
```

bsideup replied:
Thanks, the diagram really helps! Maybe let's add it to the javadoc?

simonbasle replied:
I'll add a quick SVG version for both `multicast()` and `multicastNoWarmup()`

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/Sinks.java:
```diff
@@ -0,0 +1,262 @@
+	 * {@link ScalarSink#success() empty completions} or {@link ScalarSink#error(Throwable) error}.
+	 * This completion is replayed to late subscribers.
+	 */
+	public static <T> StandaloneMonoSink<T> trigger() {
```
WDYT about `promise` or `future`? `trigger` is too verb-ish

simonbasle replied:
I don't really like `trigger` either, future and promise sound too overloaded. Maybe `completable` to echo both `CompletableFuture` and `Completable` in RxJava? By I fear this becomes overloaded too...

bsideup replied:
IIRC `Completable` in RxJava is `Publisher<Void>`. It does sound too overloaded but we're talking about method's name, not type, so _I think_ we're fine as it helps understanding the return value much better.

simonbasle replied:
ok, I'll go with `promise` actually...

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/Sinks.java:
```diff
@@ -0,0 +1,262 @@
+				if (done) {
+					return;
+				}
+				done = true;
```
Just curious, why not using CAS?

simonbasle replied:
too much on my plate to have the perfect implementation in time for M1 :(

bsideup commented on reactor-core/src/test/java/reactor/core/publisher/SinksTest.java:
```diff
@@ -0,0 +1,721 @@
+		@Test
+		void honorsMultipleSubscribersBackpressure()
+				throws InterruptedException, ExecutionException {
+			ExecutorService es = Executors.newFixedThreadPool(2);
```
Should be destroyed after the test to avoid dangling `Thread`s

simonbasle replied:
done in the dynamic test version

bsideup commented on reactor-core/src/test/java/reactor/core/publisher/SinksTest.java:
```diff
@@ -0,0 +1,721 @@
+		@Test
+		void honorsMultipleSubscribersBackpressure()
+				throws InterruptedException, ExecutionException {
+			ExecutorService es = Executors.newFixedThreadPool(2);
```
cleanup is missing

simonbasle replied:
done in the dynamic test version

bsideup commented on reactor-core/src/test/java/reactor/core/publisher/SinksTest.java:
```diff
@@ -0,0 +1,721 @@
+		}
+
+		@Test
+		void fluxViewReturnsSameInstance() {
```
could use an abstract parent class that checks common behaviours like this one

simonbasle replied:
done: made a dynamic test generation to express more directly the semantics found in the javadocs

bsideup commented on reactor-core/src/test/java/reactor/core/publisher/SinksTest.java:
```diff
@@ -0,0 +1,721 @@
+			first.assertValues(1).assertComplete();
+		}
+
+		@Test //TODO is that acceptable that EmitterProcessor doesn't replay in case of errors?
```
Probably not. Especially if `onErrorResume` is used on the returning `Flux`

simonbasle replied:
isolated that in the test a bit more, will have to decide what to do by GA

bsideup commented on reactor-core/src/test/java/reactor/core/publisher/SinksTest.java:
```diff
@@ -0,0 +1,721 @@
+		@Test
+		void honorsMultipleSubscribersBackpressure()
+				throws InterruptedException, ExecutionException {
+			ExecutorService es = Executors.newFixedThreadPool(2);
```
ditto (the cleanup)

simonbasle replied:
done in the dynamic test

simonbasle commented:
@bsideup I added more commits to adress your review:
 - Rework SinksTests to mutualize code as dynamic tests
 - improve javadocs and add svg for warmup concept
 - rename trigger to promise. add tests and fix a bug with error dropping
 - added `FluxIdentityProcessor<T>` specialization of `FluxProcessor<T,T>`

smaldini commented:
@dfeist tryNext/offer could be great with other features such as per-item context so we would have a true no-ack receipt if either downstream operator discards the data. Otherwise just on the sink it might prove confusing as only type of underlying processor has backpressure support, that'd be the only one sometimes returning true, sometimes false.
